File: src/test/java/com/raychenon/leetcode/GroupAnagramsTest.java
package com.raychenon.leetcode;

import org.junit.Assert;
import org.junit.BeforeClass;
import org.junit.Test;

import java.util.Arrays;
import java.util.List;

import static junit.framework.TestCase.assertEquals;


/**
 * User: raychenon
 * Date: 2019-09-10
 * https://leetcode.com/problems/group-anagrams/
 */
public class GroupAnagramsTest {

    private static GroupAnagrams groupAnagrams;

    @BeforeClass
    public static void setUp() {
        groupAnagrams = new GroupAnagrams();
    }

    @Test
    public void groupAnagramsTest() {
        List<List<String>> expectedResult = Arrays.asList(Arrays.asList("bat"), Arrays.asList("nat", "tan"), Arrays.asList("ate", "eat", "tea"));
        List<List<String>> result = groupAnagrams.groupAnagrams(new String[]{"eat", "tea", "tan", "ate", "nat", "bat"});

        assertEquals(expectedResult.size(), result.size());
        // to test the equality without order, assert that all elements are contained in a list and vice versa
        for (int i = 0; i < result.size(); i++) {
            Assert.assertTrue(expectedResult.get(i).containsAll(result.get(i)));
            Assert.assertTrue(result.get(i).containsAll(expectedResult.get(i)));
        }
    }

    @Test
    public void groupShortAnagramsTest() {
        List<List<String>> result = groupAnagrams.groupAnagrams(new String[]{"eat", "tea", "ate", "bat"});
        List<List<String>> expectedResult = Arrays.asList(Arrays.asList("eat", "tea", "ate"), Arrays.asList("bat"));

        Assert.assertTrue(expectedResult.containsAll(result));
    }
}

File: src/test/kotlin/com/raychenon/kotlin/leetcode/GroupAnagramsTest.kt
package com.raychenon.kotlin.leetcode

import org.junit.Assert
import org.junit.Test
import java.util.*

/**
 * User: raychenon
 * Date: 2019-09-27
 * https://leetcode.com/problems/group-anagrams/
 */
class GroupAnagramsTest {

    @Test
    fun groupAnagramsTest() {
        val expectedResult =
            Arrays.asList(Arrays.asList("ate", "eat", "tea"), Arrays.asList("bat"), Arrays.asList("nat", "tan"))
        val result = GroupAnagrams.groupAnagrams(arrayOf("eat", "tea", "tan", "ate", "nat", "bat"))

        Assert.assertEquals(expectedResult.size, result.size)
        // to test the equality without order, assert that all elements are contained in a list and vice versa
        Assert.assertTrue(expectedResult.get(0).containsAll(result.get(0)))
        Assert.assertTrue(result.get(0).containsAll(expectedResult.get(0)))

        Assert.assertTrue(expectedResult.get(1).containsAll(result.get(2)))
        Assert.assertTrue(expectedResult.get(2).containsAll(result.get(1)))
    }

    @Test
    fun groupShortAnagramsTest() {
        val result = GroupAnagrams.groupAnagrams(arrayOf("eat", "tea", "ate", "bat"))
        val expectedResult = Arrays.asList(Arrays.asList("eat", "tea", "ate"), Arrays.asList("bat"))
        Assert.assertTrue(expectedResult.containsAll(result))
    }

}